### PR TITLE
Map type config to support metrics tags

### PR DESCRIPTION
**Pull request by mageshn**

### CODE REVIEW

File: config/src/main/java/io/confluent/common/config/ConfigDef.java [hunk 1]
```diff
@@ -336,6 +336,27 @@ else if (value instanceof String)
           } else {
             throw new ConfigException(name, value, "Expected a comma separated list.");
           }
+        case MAP:
+          if (value instanceof Map) {
+            return value;
+          } else if (value instanceof String) {
+            if(trimmed.isEmpty()) {
```
Comment (edenhill) [inline config/src/main/java/io/confluent/common/config/ConfigDef.java]: inconsistent spacing, should be `if (`

File: config/src/main/java/io/confluent/common/config/ConfigDef.java [hunk 2]
```diff
@@ -336,6 +336,27 @@ else if (value instanceof String)
...
+              return Collections.emptyMap();
+            } else {
+              Map<String, String> map = new HashMap<>();
+              List<String> entries = Arrays.asList(trimmed.split("\\s*,\\s*", -1));
+              for ( String entry : entries) {
```
Comment (edenhill) [inline config/src/main/java/io/confluent/common/config/ConfigDef.java]: extra space inserted before `String`

File: config/src/main/java/io/confluent/common/config/ConfigDef.java [hunk 3]
```diff
@@ -336,6 +336,27 @@ else if (value instanceof String)
...
+                String[] keyValue = entry.split("\\s*:\\s*", -1);
```
Comment (edenhill) [inline config/src/main/java/io/confluent/common/config/ConfigDef.java]: same comment as in the other PR: "=" as delimiter seems more natural. (but I don't really have anything to back that up)

File: config/src/test/java/io/confluent/common/config/ConfigDefTest.java [hunk 4]
```diff
@@ -77,6 +79,7 @@ public void testBasicTypes() {
     props.put("h", "FalSE");
     props.put("i", "TRUE");
     props.put("j", "password");
+    props.put("k", "k1:v1,k2:v2");
```
Comment (edenhill) [inline config/src/test/java/io/confluent/common/config/ConfigDefTest.java]: add a test for single key-value (no , separator)

File: config/src/main/java/io/confluent/common/config/ConfigDef.java [hunk 5]
```diff
@@ -336,6 +336,27 @@ else if (value instanceof String)
...
+                if (keyValue.length != 2) {
+                  throw new ConfigException("Map entry should be of form <key>:<value");
```
Comment (wicknicks) [inline config/src/main/java/io/confluent/common/config/ConfigDef.java]: missing a ">" after value.

**Comment by mageshn** [comment]
@ewencp this is to provide ability to configure metrics tags. This will help add additional tags for monitoring systems. Kafka has a similar config listener.security.protocol.map but then its defined as a String instead of a Map.

File: config/src/main/java/io/confluent/common/config/ConfigDef.java (hunk 3 above)
Comment (mageshn) [reply config/src/main/java/io/confluent/common/config/ConfigDef.java]: I based it off the config in Kafka listener.security.protocol.map. The separator here is ':'

**Comment by mageshn** [comment]
@edenhill ping :)

**Pull request merged**